### DISCUSSION TITLE
mutate a new column with recurring values

### DISCUSSION THREAD

**Discussion by alexandro-ag** [markfairbanks/tidypolars#158] [Ideas]
In R, it's possible to mutate new columns with vectors of values:

``` r
library(tidyverse)

(test_df <- tibble(x = rep("a", 5), y = rep(0L, 5)))
#> # A tibble: 5 x 2
#>   x         y
#>   <chr> <int>
#> 1 a         0
#> 2 a         0
#> 3 a         0
#> 4 a         0
#> 5 a         0

test_df %>% 
  mutate(chr_col = rep("new_value", 5), ## Could specify
         int_col = rep(2L, 5))          ## length of values
#> # A tibble: 5 x 4
#>   x         y chr_col   int_col
#>   <chr> <int> <chr>       <int>
#> 1 a         0 new_value       2
#> 2 a         0 new_value       2
#> 3 a         0 new_value       2
#> 4 a         0 new_value       2
#> 5 a         0 new_value       2

test_df %>% 
  mutate(chr_col = "new_value", int_col = 2L) ## Or not
#> # A tibble: 5 x 4
#>   x         y chr_col   int_col
#>   <chr> <int> <chr>       <int>
#> 1 a         0 new_value       2
#> 2 a         0 new_value       2
#> 3 a         0 new_value       2
#> 4 a         0 new_value       2
#> 5 a         0 new_value       2
```

<sup>Created on 2021-12-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Would it be possible to replicate this behavior in tidypolars?

``` python
import tidypolars as tp
from numpy import repeat as rep

test_df = tp.Tibble(x = rep('a', 5), y = rep(0, 5))
#   x | y
#  -- | --
# str | i32
# "a" | 0
# "a" | 0
# "a" | 0
# "a" | 0
# "a" | 0


test_df.mutate(chr_col = rep('new_value', 5), int_col = rep(2, 5))
# AttributeError: 'numpy.ndarray' object has no attribute 'alias'
test_df.mutate(chr_col = 'new_value', int_col = 2)
# AttributeError: 'str' object has no attribute 'alias'
```

**Comment by markfairbanks** [comment]
If your input to `.mutate()` is a python object (string, float, integer, list) or a numpy array you either need to convert them to a `Series` using `tp.Series` or mark them as a "literal value" using `tp.lit()`.

I'll also open an issue to implement `rep()` so that you don't need numpy. (Edit: I added `rep()` to this issue: #90)

**Option 1 - Convert to a Series**

```python
test_df.mutate(chr_col = tp.Series(rep('new_value', 3)), int_col = tp.Series(rep(2, 3)))
test_df.mutate(chr_col = tp.Series(['new_value']), int_col = tp.Series([2]))
```

In polars a Series can be thought of more or less as an R vector.

**Option 2 - Mark the input as a literal value**

```python
test_df.mutate(chr_col = tp.lit(rep('new_value', 5)), int_col = tp.lit(rep(2, 5)))
test_df.mutate(chr_col = tp.lit('new_value'), int_col = tp.lit(2))
```

When using `tp.lit()` you actually create an expression object that doesn't evaluate unless inside `.mutate()`. It works somewhat like `col()`, but `col()` marks the object as a column inside the data frame. Whereas `lit()` tells polars to look for the value itself.

Note: `lit()` is only necessary if you aren't using `col()` inside of your mutate expression!

Hope this helps - if I need to clarify anything let me know.

**Reply by alexandro-ag** [reply]
this helps a lot! thank you for the explanation.

i tried 
``` python
test_df.mutate(chr_col = tp.lit('new_value'), int_col = tp.lit(2))
```
and the solution works for me.

**Comment by markfairbanks** [comment]
FYI @alexandro-ag - in the latest version of tidypolars you can now add constant columns without using `tp.lit()`:

```python
import tidypolars as tp
from tidypolars import col

df = tp.Tibble(x = range(3), y = range(3))

df.mutate(z = "z")
```
```
┌─────┬─────┬─────┐
│ x   ┆ y   ┆ z   │
│ --- ┆ --- ┆ --- │
│ i64 ┆ i64 ┆ str │
╞═════╪═════╪═════╡
│ 0   ┆ 0   ┆ z   │
├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤
│ 1   ┆ 1   ┆ z   │
├╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌┤
│ 2   ┆ 2   ┆ z   │
└─────┴─────┴─────┘
```

**Comment by alexandro-ag** [comment]
this is fantastic! thank you so much for the heads up and hard work!!
